### PR TITLE
fix: better content negotiation

### PR DESCRIPTION
**Pull request by vwong**

### CODE REVIEW

File: lib/swagger-mock-validator/validate-spec-and-mock/content-negotiation.ts [hunk 1]
```diff
@@ -1,49 +1,90 @@
...
+const type = (t: string): string => t.split(TYPE_SUBTYPE_SEPARATOR)[0];
+const subtype = (t: string): string | undefined => {
+    const [_type, subtype] = t.split(TYPE_SUBTYPE_SEPARATOR);
+    return subtype?.split(EXTENSION_SEPARATOR).pop();
 };
 
-export const areMediaTypesCompatible = (actualMediaType: string, supportedMediaType: string): boolean => {
-    const parsedActualMediaType = parseMediaType(actualMediaType);
-    const parsedSupportedMediaType = parseMediaType(supportedMediaType);
+export function findMatchingType(requestType: string, responseTypes: string[]): string | undefined {
```
Comment (vwong) [inline lib/swagger-mock-validator/validate-spec-and-mock/content-negotiation.ts:24]: This is the function of interest

File: test/unit/content-negotiation.spec.ts [hunk 2]
```diff
@@ -1,126 +1,234 @@
...
-        const result = isMediaTypeSupported('text/plain/html', ['text/plain/css']);
+        const result = findMatchingType('text/plain/html', ['text/plain/css']);
 
-        expect(result).toBe(true);
+        expect(result).toBeTruthy();
+    });
+});
+
+// more specific tests
+describe('#findMatchingType', () => {
```
Comment (vwong) [inline test/unit/content-negotiation.spec.ts:130]: But here's the new more specific behaviour.

File: lib/swagger-mock-validator/validate-spec-and-mock/validate-parsed-mock-response-body.ts [hunk 3]
```diff
@@ -68,7 +68,7 @@ export const validateParsedMockResponseBody = (
     return [];
   }
 
-  const expectedMediaType = parsedMockInteraction.requestHeaders.accept?.value;
+  const expectedMediaType = parsedMockInteraction.requestHeaders.accept?.value || "application/json";
```
Comment (mefellows) [inline lib/swagger-mock-validator/validate-spec-and-mock/validate-parsed-mock-response-body.ts:71]: Is the idea here that we default to JSON in the case no accept header is defined or found?

File: test/unit/content-negotiation.spec.ts [hunk 4]
```diff
@@ -1,126 +1,233 @@
...
+        it('matches exactly', () => {
+            expect(findMatchingType('application/vnd.foo+json', ['application/json', 'application/vnd.foo+json'])).toBe(
+                'application/vnd.foo+json'
+            );
+            expect(findMatchingType('application/json', ['application/json', 'application/vnd.foo+json'])).toBe(
+                'application/json'
+            );
+        });
+
+        it('fallsback to base content-type', () => {
```
Comment (mefellows) [inline test/unit/content-negotiation.spec.ts:182]:
```suggestion
        it('falls back to base content-type', () => {
```

File: test/unit/content-negotiation.spec.ts [hunk 5]
```diff
@@ -1,126 +1,233 @@
...
+            expect(findMatchingType('application/vnd.foo+json', ['application/json'])).toBe('application/json');
```
Comment (mefellows) [inline test/unit/content-negotiation.spec.ts:183]: I think this answers the above.

File: test/unit/content-negotiation.spec.ts [hunk 6]
```diff
@@ -1,126 +1,233 @@
...
+        });
+
+        it('uses extended content-type', () => {
+            expect(findMatchingType('application/json', ['application/vnd.foo+json'])).toBe('application/vnd.foo+json');
+        });
+    });
+
+    describe('wildcard responses', () => {
+        it('matches wildcard types', () => {
+            expect(findMatchingType('aaa', ['aaa/*'])).toBe('aaa/*');
```
Comment (mefellows) [inline test/unit/content-negotiation.spec.ts:193]: Is this even a thing? What would it mean for an OAS to say I return a wild card subtype?

**Review by mefellows — APPROVED** [review]
Looks good - only minor questions. Do you need to consider adding any of the case sensitivity tests to the new bits? I saw one earlier, which might suffice.

File: test/unit/content-negotiation.spec.ts (hunk 6 above)
Comment (vwong) [reply test/unit/content-negotiation.spec.ts:193]: I've seen examples of this, where the OAS mentions things beyond the type/subtype (eg. `application/json; charset=utf-8`). For completeness, I assume it makes sense to allow the OAS to specify `image/*` meaning the provider can return with any image type.

File: lib/swagger-mock-validator/validate-spec-and-mock/validate-parsed-mock-response-body.ts (hunk 3 above)
Comment (vwong) [reply lib/swagger-mock-validator/validate-spec-and-mock/validate-parsed-mock-response-body.ts:71]: Yes. Previously this defaulting happened elsewhere. But I've relocated it here.

File: test/unit/content-negotiation.spec.ts (hunk 5 above)
Comment (vwong) [reply test/unit/content-negotiation.spec.ts:183]: Not quite. This means that it strips the vendor extension.

Comment (mefellows) [reply test/unit/content-negotiation.spec.ts:183]: Ah yes, got it.

File: test/unit/content-negotiation.spec.ts (hunk 6 above)
Comment (mefellows) [reply test/unit/content-negotiation.spec.ts:193]: 👉 https://swagger.io/docs/specification/media-types/

Screenshot 2023-05-24 at 2 56 24 pm

So, best you keep the test ;)

**Comment by vwong** [comment]
@mefellows Added a couple case-sensitivity tests

**Comment by mefellows** [comment]
> @mefellows Added a couple case-sensitivity tests

Thanks!

**Pull request merged**